### DISCUSSION TITLE
shapeDat

### DISCUSSION THREAD

**Discussion by jjburton** [jjburton/cgmTools#257] [Ideas]
Working through a new ShapeDat setup to be able to transfer shapes between assets.

- relative to root

There's a number of things we need to gather up.

- [x] Get form handles dat
- [ ] Settings
- [ ] Handle lever

**Comment by jjburton** [comment]
### First pass

```python
import cgm.core.mrs.ShapeDat as SHAPEDAT
reload(SHAPEDAT)

mBlockTarget = cgmMeta.asMeta('staff_handle_shaperList_handleBlock')
sDat = SHAPEDAT.dat_get(mBlock)
SHAPEDAT.dat_set(mBlockTarget, sDat)

pprint.pprint(sDat)

mSource = cgmMeta.asMeta('staff_limb_2_1_shapeHandle')
mTar = cgmMeta.asMeta('staff_limb_2_0_shapeHandle')


dat = SHAPEDAT.shapes_get(mSource)
dat['os']

SHAPEDAT.shapes_set(mTar,dat)
```

**Reply by jjburton** [reply]
### Night 2

- Form, Loft, Shape loading working
- Tested with Json. Must use string keys so retooled data from nested indice dict to just nested list

```python

with open("D:\\New Folder\\test.json", "w") as fp:
	json.dump(sDat, fp, indent=1)


#%%
with open("D:\\New Folder\\test.json", "r") as fp:
	loaded_data = json.load(fp)
```

**Comment by jjburton** [comment]
### Loading to new
- Save local shapeDat
- Build a new handle, limb, segment, head(?) with that shape dat
- Test with/without lever
- Account for too few bits of data and too much

- [ ] Handle
- [ ] Limb
- [ ] Segment
- [ ] Head

**Comment by jjburton** [comment]
### How to store a relative point of data
The continual problem I've had with trying to figure out how to store and load block dat is one of consistency. Matrix or relative matrix data can introduce shear and I've been unable to get consistent results because matrices require replicating the same heirarchy for consistent results. As such I've been noodling with how to consistently store a given point relative to a block root for some time.

My current thinking is the best way will be:
- the block root is the given known
- shapeDirection or an assumed vector of that node will give me a consistent base vector
- taking a given control or point I want to know and getting the vector differential from the root to that point and distance will give me consistent value
- I'll probably need to account distance to blockScale though can probably work around that

**Reply by jjburton** [reply]
This was helpful - https://stackoverflow.com/questions/47292653/get-an-objects-position-in-another-objects-coordinate-system

**Reply by jjburton** [reply]
Got relative position working though I still think the vector distance setup is going to be more consistent.

Relative orientation is more fickle, rotation order/etc and all that. If I can get base vectors consistently transformed from the block root I should in theory be able to reconstruct orientation consistently by a two vector aim, maybe three. 

Gonna ping a few friends and loop back

**Comment by jjburton** [comment]
### 7.13
Looping back. After talking to some mates the most straight forward solution is to to just move and zero the block when saving/loading data.

- [x] convert save to zero block method
- [x] convert load to zero block method
- [x] add cvs only load

positional info for shape handles isn't quite what I want. Translation isn't ideal either....how do we best define a point relative to other points. A bounding box setup might work or a scaleSpace like on face blocks though I'm not sure of a super efficient way to establish that scale space.

Toying with a method of defining that space via a point, some vectors and distance values.
![image](https://user-images.githubusercontent.com/48218358/125635343-34ac105d-1306-4c93-9f58-1b0e2757bcf7.png)

**Comment by jjburton** [comment]
### 7.20
Wanna try this relative setup...

```python
import cgm.core.mrs.ShapeDat as SHAPEDAT
reload(SHAPEDAT)

_target = 'relative_target'
_start = 'relative_start'
_end = 'relative_end'
d = SHAPEDAT.relativePointDat_getFromObjs(_target,_start,_end)

_targetMove = 'relative_target2'
SHAPEDAT.relativePointDat_setFromObjs(_targetMove,_start,_end, d)
```

**Reply by jjburton** [reply]
### 7.26

Working on limb,segment,handle initially
share.rd.scenes.staff

```python
import cgm.core.mrs.ShapeDat as SHAPEDAT
reload(SHAPEDAT)
mc.ls(sl=1)
_targets = [u'relative_target3', u'relative_target2', u'relative_target', u'relative_target1']
_target = 'relative_target'
_start = 'relative_start'
_end = 'relative_end'
d = SHAPEDAT.relativePointDat_getFromObjs(_targets,_start,_end)

_targetMove = 'relative_target2'
SHAPEDAT.relativePointDat_setFromObjs(_targets,_start,_end, d)

mBlock.select()
mBlockTarget = cgmMeta.asMeta('staff_segment_simple_segmentBlock')
mBlockTarget = cgmMeta.asMeta('box__handleBlock')
mBlockTarget = cgmMeta.asMeta('staff_handle_shaperList_handleBlock')


mBlockTarget.p_position = 0,0,0
mBlock.blockState
reload(SHAPEDAT)
sDat = SHAPEDAT.dat_get(mBlock)
mBlockTarget.select()
SHAPEDAT.dat_set(mBlockTarget, sDat, formHandles = 1, loftHandles = 1, loftShapes = 1)
SHAPEDAT.dat_set(mBlockTarget, sDat, formHandles = 0, loftHandles = 1, loftShapes = 0)
SHAPEDAT.dat_set(mBlockTarget, sDat, formHandles = 0, loftHandles = 0, loftShapes = 1)
SHAPEDAT.dat_set(mBlockTarget, sDat, formHandles = 0, loftHandles = 1)
SHAPEDAT.dat_set(mBlockTarget, sDat, formHandles = 1, loftHandles = 0)

pprint.pprint(sDat)
pprint.pprint(sDat['settings'])
pprint.pprint(sDat['base'])

pprint.pprint(sDat['handles'].keys())
pprint.pprint(sDat['handles']['form'])
pprint.pprint(sDat['handles']['shapes'])
pprint.pprint(sDat['handles']['loft'])
pprint.pprint(sDat['handles']['loftShapes'])
pprint.pprint(sDat['handles']['sub'])
pprint.pprint(sDat['handles']['subShapes'])
pprint.pprint(sDat['handles']['subRelative'][0])
```

**Comment by benngarnish** [comment]
Continuing to the conversation from discord here:

male and female needs enough form handles and sub handles to indicate the chest area. I find the female characters need more for the bust. 
mad you have the rest of the form handles already give enough details for me.

**Reply by jjburton** [reply]
Was baseFemale sufficient or some changes there?

**Comment by jjburton** [comment]
### Weekend 7.31
Setting up first ui...
![image](https://user-images.githubusercontent.com/48218358/127749865-93f70201-9198-45ac-be9b-3ac765c561dc.png)

- [x] frame
- [x] data section frame
- [x] check boxes
- [x] query
- [x] load

Cases
- [x] no lever to lever (thumb to finger)
- [x] finger to thumb

ui work today
![image](https://user-images.githubusercontent.com/48218358/127757164-dabeb6b9-61bf-4dd5-8490-e919c6b79cad.png)

**Comment by jjburton** [comment]
```POST.swimSettings_get('tailSwimautoSwim_wave_settings_grp','tailSwimautoSwim_wave_handle')```
